internal/controller/controllers/crd_utils.go

```diff
@@ -50,15 +58,35 @@ func (u *CRDUtils) CreateAgentCR(ctx context.Context, log logrus.FieldLogger, ho
 			},
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      hostId,
-				Namespace: clusterNamespace,
+				Namespace: infraEnv.Namespace,
 			},
 		}
-		log.Infof("Creating Agent CR. Namespace: %s, Cluster: %s, HostID: %s", clusterNamespace, clusterName, hostId)
+		log.Infof("Creating Agent CR. Namespace: %s, Cluster: %s, HostID: %s", infraEnv.Namespace, clusterName, hostId)
 		return u.client.Create(ctx, host)
 	}
 	return err
 }
 
+// Returns the InfraEnv resource associated with the cluster
+func (u *CRDUtils) getInfraEnvByCluster(ctx context.Context, log logrus.FieldLogger, clusterNamespace, clusterName string) (*aiv1beta1.InfraEnv, error) {
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	deploymentNamespacedName := types.NamespacedName{
+		Name:      clusterName,
+		Namespace: clusterNamespace,
+	}
+	err := u.client.Get(ctx, deploymentNamespacedName, clusterDeployment)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("Failed to get ClusterDeployment resource %s", deploymentNamespacedName))
+	}
+
+	infraEnv, err := getInfraEnvByClusterDeployment(ctx, u.client, clusterDeployment)
```

Comment:
You should also check that `infraEnv` is not `nil`

the function `getInfraEnvByClusterDeployment()` will not return an error in such case. It uses `getInfraEnvByClusterDeployment()` internally which will return `nil, nil` of there is no `InfraEnv` to be found for the `clusterDeployment`

https://github.com/openshift/assisted-service/blob/f4fc64b827be1213f7c98a9c4b207c0aaf1de32b/internal/controller/controllers/common.go#L44-L57

This may lead to a nil pointer exception on line#61

Reply:
while it is true that users should create `InfraEnv` to get the image for registering the hosts, we could theoretically see a case in which a user generated and downloaded the image, and deleted InfraEnv.

Kind of a long shot, but still worth not having a nil pointer for it

Reply:
Expect the unexpected. :)

Reply:
It's a good question to ask what should be done in this case. 
We probably should return an error and not allow that agent to register.

Reply:
I agree.
We should also log this (perhaps with a warning?)

Reply:
Makes sense. Handled missing InfraEnv resource scenario and added relevant logs and tests.